juju/juju
# Fix action CLI display issue (ordering of tasks)

The juju run CLI commands (actions and exec commands) were still using the over-the-wire params structs when calling the client side API. A side effect of this was that operation IDs were handled as strings (from tags) and so ordering was messed up. For numeric IDs we want natural sort order but we were getting straight alphabetical.

A large part of the fix here is to provide new bespoke structs for the api layer to use with the CLI. These structs use IDs not tags.
Along with the structs, a bunch of unmarshalling helper functions are provided. There are some deprecated structs and methods (already removed in 3.0) used with the legacy action UUIDs which still use params structs in the 2.9 branch.

The functional change is to rely on the fact that there are now IDs and perform a natural sort of the operations.

Tests were updated not to use JujuConnSuite (nor mongo) anymore (use the sub api caller pattern instead).
Plus extra (missing) tests were added for the Run() APIs.

## QA steps

deploy a charm with actions
```
juju deploy foo -n 2
juju exec --machine 0 "pwd"
juju exec --machine 0 "pwd" --background
juju exec --application ubuntu-lite "pwd"
```
exec several more commands to get the operation count > 0

check UX messages for wording missteps
```
juju operations --machines 0
juju operations
```
check ordering and correct display of task IDs, summary, receive names etc
```
juju show-task X --format yaml
```
now run some actions
```
juju run ubuntu-lite/0 hello
juju run ubuntu-lite/0 hello --background
```
again use list-operations and show-task to check output
